### PR TITLE
Remove component name detection

### PR DESCRIPTION
**Pull request by sevos**
Please revie #14 first.

Remove automatic name detection, allow passing many components to WebpackerReact.register

Name detection did not work for functional components. Passing an object with component names as keys and components as values allows to leverage ES6 shorthand for objects (see changes in the README.md).

Please look at the second commit to see changes. I based changes here on my previous pull request (sorry for that - I can clean this up later).

### CODE REVIEW

File: javascript/webpacker_react-npm-module/package.json [hunk 1]
```diff
@@ -7,5 +7,8 @@
   "devDependencies": {
     "babel-cli": "^6.18.0",
     "babel-preset-es2015": "^6.18.0"
+  },
+  "dependencies": {
+    "lodash": "^4.17.4"
```
Comment (sevos) [inline javascript/webpacker_react-npm-module/package.json]: I should probably loosen this dep

File: javascript/webpacker_react-npm-module/src/index.js [hunk 2]
```diff
@@ -119,16 +120,15 @@ var WebpackerReact = {
...
   },
 
-  register: function (component) {
-    var name = component.name
-
-    if (this.registeredComponents[name]) {
-      console.warn('webpacker-react: Cant register component, another one with this name is already registered: ' + name)
-      return false
+  register: function (components) {
+    var collisions = _.intersection(_.keys(this.registeredComponents), _.keys(components));
```
Comment (sevos) [inline javascript/webpacker_react-npm-module/src/index.js]: Should we really check against collisions? I think it's overprotective and usually there will be one place in the app for registering components.

File: test/example_app/app/javascript/packs/hello_react.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import Hello from 'components/hello';
+import HelloReact from 'components/hello';
 import WebpackerReact from 'webpacker-react';
 
-WebpackerReact.register(Hello);
+WebpackerReact.register({HelloReact});
```
Comment (sevos) [inline test/example_app/app/javascript/packs/hello_react.js]: I changed `Hello` to `HelloReact` to point out that the class name inside the component does not matter.

**Comment by renchap** [comment]
Can you rebase it to the latest master and fix the linting, then squash?

I am not sure if we should drop support for the current syntax, or keep both at the same time.

**Comment by sevos** [comment]
Done :D

**Comment by sevos** [comment]
The problem is that current syntax can not work with functional components, and when you start using redux or other libraries, your components are wrapped with functional components

File: README.md [hunk 4]
```diff
@@ -50,8 +50,12 @@ In your pack file (`app/javascript/packs/*.js`), import your components as well
 import Hello from 'components/hello'
 import WebpackerReact from 'webpacker-react'
 
+<<<<<<< HEAD
```
Comment (renchap) [inline README.md]: @sevos You messed up your merge here :)

**Comment by renchap** [comment]
I definitely agree with your proposal, I am just wondering if we should keep the simpler `WebpackerReact.register(HelloReact, HellaReact)` syntax.

Also, you should update the README with an example for multiple components.

**Comment by sevos** [comment]
There are three questions to answer:

1. How to register components? I have two ideas:
  * Allow single `initialize` call which will take the object `{name: component, ...}`. This would enforce convention of single components manifest per pack
  * `initialize` is called in the pack file and components are registered by high order function within the components file
    ```javascript
    import {unobtrusiveComponent} from 'webpacker-react'

    ...

    export default unobtrusiveComponent({Hello})
    ```

2. Should every pack register own Webpacker::React handlers, or should Webpacker::React be a singleton, like Turbolinks, and be mounted in `window`? *IMHO: singleton*

3. If there are many packs: should components be under single namespace from the perspective of Rails' helper, or should be namespace them? In that case `initialize` would take the name of the namespace. *IMHO: for now just single, global namespace with collision prevention*

**Comment by sevos** [comment]
@renchap if we keep the old syntax, how do you want to prevent the component's name to be empty, when a functional component is passed?

**Comment by sevos** [comment]
We could deprecate the old syntax, encouraging switching to the new and throw an error if the name can not be determined.

File: README.md (hunk 4 above)
Comment (sevos) [inline README.md]: fixed, thx

**Comment by sevos** [comment]
BTW: how these changes play with HMR?

**Comment by renchap** [comment]
> Allow single initialize call which will take the object {name: component, ...}. This would enforce convention of single components manifest per pack

This seems a good alternative, similar to having `initialize` ran on the first `register` call. I think I prefer the function to be called `register` and have a side effect of initialising the callbacks.

> Should every pack register own Webpacker::React handlers, or should Webpacker::React be a singleton, like Turbolinks, and be mounted in window? IMHO: *singleton*

Definitely singleton. Is it a way to avoid setting a global/window variable?

> If there are many packs: should components be under single namespace from the perspective of Rails' helper, or should we put them under separate namespaces? In that case initialize would take the name of the namespace. IMHO: for now just single, global namespace with collision prevention

I prefer 1 namespace too.

> We could deprecate the old syntax, encouraging switching to the new and throw an error if the name can not be determined.

Yes, my question is: do want to depreciate the old syntax, or just keep it and throw an error if we cant get the name. People may not be familiar with the `{Object}` syntax and may be confused when reading the doc, but I guess it can be explained (and put a comment like "shorthand for `{ object: object }`

**Comment by sevos** [comment]
https://github.com/sevos/webpacker-react/pull/1 Please take a look @renchap

**Comment by sevos** [comment]
`{Object}` syntax is a common practice once you get into React and Redux, so people should be familiar with this ES6 feature. And we will document it, right? :)

**Comment by sevos** [comment]
Bump

File: javascript/webpacker_react-npm-module/package.json [hunk 5]
```diff
@@ -13,5 +13,8 @@
     "eslint-plugin-import": "^2.2.0",
     "eslint-plugin-jsx-a11y": "^4.0.0",
     "eslint-plugin-react": "^6.10.0"
+  },
+  "dependencies": {
+    "lodash": "^4.0.0"
```
Comment (renchap) [inline javascript/webpacker_react-npm-module/package.json]: I would prefer to not depend on lodash, especially if we only use 1 or 2 function

File: README.md [hunk 6]
```diff
@@ -50,8 +50,8 @@ In your pack file (`app/javascript/packs/*.js`), import your components as well
 import Hello from 'components/hello'
 import WebpackerReact from 'webpacker-react'
 
-WebpackerReact.register(Hello)
 WebpackerReact.initialize()
```
Comment (renchap) [inline README.md]: Should be `setup`

File: README.md [hunk 7]
```diff
@@ -68,7 +68,7 @@ import Turbolinks from 'turbolinks'
 Turbolinks.start()
 WebpackerReact.initialize()
```
Comment (renchap) [inline README.md]: Same

File: javascript/webpacker_react-npm-module/src/index.js [hunk 8]
```diff
@@ -1,12 +1,12 @@
 import React from 'react'
 import ReactDOM from 'react-dom'
+import _ from 'lodash'
```
Comment (renchap) [inline javascript/webpacker_react-npm-module/src/index.js]: If we use lodash, then we should only import the functions we use, not the whole library

File: javascript/webpacker_react-npm-module/package.json (hunk 5 above)
Comment (sevos) [inline javascript/webpacker_react-npm-module/package.json]: Currently we use 4 functions, where `keys()` and `assign()` would require polyfill (as far as I understand) anyway and `intersection()` and `omit()` may require own implementation. I am happy to explore this as soon as we get the basic functionalities in place, so we can promote the gem and spread the word.

**Comment by renchap** [comment]
I just tried the current (0.1.0) implementation with a production build and the Uglify Webpack plugin and it fails because the component name is mangled by Uglify (`Root => t` for example), so we guess the name as `t`, but the attribute is still named `Root` in the HTML.
I guess this proves my initial implementation can not work 🙂

I will test your implementation and check some edge cases as well as a production build, but it should work.

I am considering dropping the old `register` API as it will break production builds and should not be used.

**Comment by sevos** [comment]
I have a test and fix for multiple pack files.

**Comment by renchap** [comment]
Looks good! Thanks for the test.
I dont really like polluting `window` but I dont think we have choice here if we have 2 different packs.I tried to implement a singleton by exporting an instance of WebpackerReact (https://k94n.com/es6-modules-single-instance-pattern), but it can't work because the 2 packs don't share anything.
I will try to release 0.2.0 tomorrow.

**Comment by sevos** [comment]
Do you think we can have release soon? Currently it's impossible to use the master branch via github (due to package.json file placement).

**Comment by renchap** [comment]
It is now released. I will try to find a way to allow to use the git repository in `package.json` to use the master version as dependancy (see #25)
